New feature: follow-mode

schwupp opened the pull request:
At first there is the fundamental reason for tping: To only monitor the change of up/down status - at lowest cost of toolsets. Or spoken differently as copied from description:
```
What you lose, is the continous reading of the RTT (you only get the first one)
What you win is a clear, timestamped view when and how long a target host went off or online
```
I thought about combining both. Dominating user experience must be the clear view of up/down events, but maybe we can put some cream on top of it.

My idea was to create an extra "status"-line at bottom, clearing it every time tping pings. Here are some questions about my first implementation:

1. only way i found so far was using 'tput' with string output paramerers (list [here](https://programmer.group/detailed-explanation-of-linux-tput-command.html)). the perfect match parameter 'ell' did somehow not work, so i workarounded with 'sc', 'rc' and 'el'. 
2. working with tput 'el' (or possibly also 'ell') has issues, when console windows forces a line break, then my code does not work. i added some more debug, so try with -d to get a longer line.
3. i think it's a big plus to see actual values of rtt, but does it distort too much to enable it per default? if we use a new switch, i would expect -f for follow, as this is pretty common elsewhere. But then we have to rename the fuzzy-feature (i'm not too proud of this name...) or we activate per default and create a new switch (i.e. -s for wellknown "static"-mode)
4. Try follow mode with fuzzy mode, i think the combination of both in action looks quite nice, you can see if fuzzy is in action now.

Biggest changes in code are:

- adding a third clause in each host-up and host-down main-loop. this also made some statistic things easier and looks better in general i think
- adding some 'tput' lines in each clause and before the main loop

schwupp commented:
@lippl can you have a look at this branch?

lippl commented:
Looks good to me!
Regarding your questions:
2. I am experiencing no issues with this on my Ubuntu wsl via vscode
3. I am perfectly fine having this as default and do not see a reason for the static otion unless you want to mitigate possible compatibility issues as mentioned in 2.
4. do you mean the initial note printed or should there be extra info in the status line? Might be a good addition for the debug mode, to display fuzzy counts. Naming wise you could rename it to reliability or interrupt mode/count!?

As the changes are breaking to #9, you might want to go ahead merging this and I will add the logic again afterwards, or you merge them by yourself. Just let me know and put your thoughts on that PR as well.

Once done, let me test on macOS too, before merging

schwupp commented:
2. i can also not recreate the issue. the easiest way to use a small console window and then start tping so it is forced to do a line break. but it seems to work for me now.

3. without (2) in effect, i will make follow mode default and implement a "-s" switch for legacy static mode

4. when tping is successful and target goes offline, fuzzy logic will keep host up until fuzzy timer finishes. this state is printed inline:

```
$ ./tping.sh -f10 192.168.1.80

Note: fuzzy dead-detection in effect, will ignore up to 10 failed pings (you will see a --FUZZY-- indicator if in action). Use for unreliable connections only.

2023-05-13 16:11:08 | host 192.168.1.80 () is ok | RTT 1.16ms
2023-05-13 16:11:18 | host 192.168.1.80 () is ok for 10 sec | RTT 0.752ms
```
ping target goes offline, fuzzy kicks in
```
2023-05-13 16:11:08 | host 192.168.1.80 () is ok | RTT 1.16ms
2023-05-13 16:11:38 | host 192.168.1.80 () is ok for 30 sec | RTT 0.762ms --FUZZY--
```
ping target still offline, fuzzy times out
```
2023-05-13 16:11:08 | host 192.168.1.80 () is ok | RTT 1.16ms
2023-05-13 16:11:50 | host 192.168.1.80 () is down [ok for 42 sec]

```

i will release some minor versions before 6.0 including #6 and #8. would be nice if you then add your fix for #7 again afterwards. then we do some testing again and finally the merge to v6.0

schwupp commented:
merging was not that easy as expected ;-) let's do extensive testing now, don't know if i broke something while merging

schwupp commented:
reverted to the state before merging #9 as this was causing a mess. as #8 was non-breaking, i already included it in main and merged in back in here. so we should have a new clean workbase for the new follow-feature

schwupp commented:
@lippl : added you as maintainer so that we can work together in this PR, if you're ok with that?

schwupp commented:
> I actually meant to merge this PR here first and then merging my other commits. As this turned into a bit of a mess, and I am not able to contribute to your feature branch anyways, I took the liberty in rewriting the feature in my repo on top of the current main branch. I also added some more fuzzy statistics there too: [feat-follow-mode](https://github.com/lippl/tping/tree/feat-follow-mode) Have a look and let me know if you would like me to just raise a PR with that branch and we close this one down? It is just missing the new version number for you to add ;)

new stats are looking quite nice!

while re-merging main, i also reverted to v5.1. i was in doubt about changing version number at the beginning or at the end of PR. right now it's the current version from main and if we decide to merge, we will raise version as last commit
